google-cloud-spanner/lib/google/cloud/spanner/interval.rb

```diff
@@ -0,0 +1,176 @@
+# Copyright 2017 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License")
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Google
+  module Cloud
+    module Spanner
+      class Interval
+        NANOSECONDS_IN_A_SECOND = 1000000000
+        NANOSECONDS_IN_A_MINUTE = NANOSECONDS_IN_A_SECOND * 60
+        NANOSECONDS_IN_AN_HOUR = NANOSECONDS_IN_A_MINUTE * 60
+        YEARS = 0
+        MONTHS = 1
+        DAYS = 2
+        HOURS = 4
+        MINUTES = 5
+        SECONDS = 6
+
+        def self.fromIso8601 interval_string
+          pattern = /
+            (?!$)(-?\d+Y)?(-?\d+M)?(-?\d+D)?(T(?=-?.?\d)(-?\d+H)?(-?\d+M)?(-?(((\d*)((\.|,)\d{1,9})?)|(\.\d{1,9}))S)?)?$
+            /
+          interval_months = 0
+          interval_days = 0
+          interval_nanoseconds = 0
+
+          captures = interval_string.match(pattern).captures
+          if captures.empty?
+            raise ArgumentError, "The ISO8601 provided was not in the correct format"
+          end
+
+          filter = Set[YEARS, MONTHS, DAYS, HOURS, MINUTES, SECONDS]
+          captures.each_with_index do |value, index|
```

Comment:
It's kind of awkward to have lots of indexed (numeric) captures. Having the indexes as constants, separated from the definition of the pattern, couples those parts of the code together and invites bugs if the pattern gets edited without also editing the constants.

It would probably be better to use named captures. (https://ruby-doc.org/3.4.1/Regexp.html#class-Regexp-label-Named+Captures). Simple example:

```ruby
pattern = /(?<year>-?\d+Y)?(?<month>-?\d+M)?(?<day>-?\d+D)?/
match = pattern.match "2025Y03M"
assert_equal "2025Y", match[:year]
assert_equal "03M", match[:month]
assert_nil match[:day]
```